sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/ShufflePartitionsUtil.scala

```diff
@@ -61,7 +63,9 @@ object ShufflePartitionsUtil extends Logging {
     // `maxTargetSize` from being set to 0.
     val maxTargetSize = math.max(
       math.ceil(totalPostShuffleInputSize / minNumPartitions.toDouble).toLong, 16)
-    val targetSize = math.min(maxTargetSize, advisoryTargetSize)
+    // The target size can be very small if minNumPartitions is super big. Here we use
+    // minPartitionSize as the lower bound to avoid too small partitions after coalescing.
+    val targetSize = math.max(math.min(maxTargetSize, advisoryTargetSize), minPartitionSize)
```

Comment:
minPartitionSize takes precedere of advisoryTargetSize? how about minPartitionSize is much greater than advisoryTargetSize?